File: LeetCode/0783. Minimum Distance Between BST Nodes/Solution.kt
package algorithm._783_Minimum_Distance_Between_BST_Nodes

/**
 * Created by Inno Fang on 2018/2/11.
 */

class TreeNode(var `val`: Int = 0) {
    var left: TreeNode? = null
    var right: TreeNode? = null
}

class Solution {

    var arr = ArrayList<Int>()
    fun minDiffInBST(root: TreeNode?): Int {
        inOrderedTraversal(root)
        var min = Int.MAX_VALUE
        for (i in 1 until arr.size) {
            min = minOf(min, arr[i] - arr[i-1])
        }
        return min
    }

    private fun inOrderedTraversal(node: TreeNode?) {
        if (node != null) {
            inOrderedTraversal(node.left)
            arr.add(node.`val`)
            inOrderedTraversal(node.right)
        }
    }
}

fun main(args: Array<String>) {
    val root = TreeNode(4)
    val node1 = TreeNode(2)
    val node2 = TreeNode(6)
    val node3 = TreeNode(1)
    val node4 = TreeNode(3)

    root.left = node1
    root.right = node2
    node1.left = node3
    node1.right = node4
    Solution().minDiffInBST(root).let(::println)
}

File: LeetCode/0783. Minimum Distance Between BST Nodes/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
 
/**
 * 46 / 46 test cases passed.
 * Runtime: 0 ms
 * Memory Usage: 9.4 MB 
 */
class Solution {
public:
    void inorder(TreeNode* root, int& pre, int& ans) {
        if (nullptr == root) return;

        inorder(root->left, pre, ans);
        if (-1 != pre) {
            ans = min(ans, abs(pre - root->val));
        }
        pre = root->val;
        inorder(root->right, pre, ans);
    }
    int minDiffInBST(TreeNode* root) {
        int ans = INT_MAX, pre = -1;
        inorder(root, pre, ans);
        return ans;
    }
};

